SonarQube.TeamBuild.PreProcessor/Roslyn/RoslynAnalyzerProvider.cs

```diff
@@ -0,0 +1,191 @@
+        /// i.e. creates the Roslyn ruleset and provisions the analyzer assemblies
+        /// and rule parameter files
+        /// </summary>
+        public static CompilerAnalyzerConfig SetupAnalyzers(ISonarQubeServer server, TeamBuildSettings settings, string projectKey, ILogger logger)
+        {
+            if (server == null)
+            {
+                throw new ArgumentNullException("server");
+            }
+            if (settings == null)
+            {
+                throw new ArgumentNullException("settings");
+            }
+            if (string.IsNullOrWhiteSpace(projectKey))
+            {
+                throw new ArgumentNullException("projectKey");
+            }
+            if (logger == null)
+            {
+                throw new ArgumentNullException("logger");
+            }
+
+            if (IsCSharpPluginInstalled(server))
+            {
+                RoslynAnalyzerProvider provider = new RoslynAnalyzerProvider(server, settings, projectKey, logger);
+                return provider.GetCompilerConfig();
+            }
+            else
+            {
+                logger.LogDebug(Resources.SLAP_CSharpPluginNotInstalled);
+            }
+
+            return null;
+        }
+
+        #endregion
+
+        #region Private methods
+
+        private static bool IsCSharpPluginInstalled(ISonarQubeServer server)
+        {
+            return server.GetInstalledPlugins().Contains(CSharpPluginKey);
+        }
+
+        private RoslynAnalyzerProvider(ISonarQubeServer server, TeamBuildSettings settings, string projectKey, ILogger logger)
+        {
+            this.server = server;
+            this.settings = settings;
+            this.projectKey = projectKey;
+            this.logger = logger;
+        }
+
+        private CompilerAnalyzerConfig GetCompilerConfig()
+        {
+            CompilerAnalyzerConfig compilerConfig = null;
+
+            RoslynExportProfile profile = TryGetRoslynConfigForProject();
+            if (profile != null)
+            {
+                compilerConfig = ProcessProfile(profile);
+            }
+            return compilerConfig;
+        }
+
+        private RoslynExportProfile TryGetRoslynConfigForProject()
+        {
+            string qualityProfile;
+            if (!this.server.TryGetQualityProfile(projectKey, CSharpLanguage, out qualityProfile))
+            {
+                this.logger.LogDebug(Resources.SLAP_NoProfileForProject, this.projectKey);
+                return null;
+            }
+
+            string profileContent = null;
+            if (!server.TryGetProfileExport(qualityProfile, CSharpLanguage, RoslynCSharpFormatName, out profileContent))
+            {
+                this.logger.LogDebug(Resources.SLAP_ProfileExportNotFound, RoslynCSharpFormatName, this.projectKey);
+                return null;
+            }
+            this.logger.LogDebug(Resources.SLAP_ProfileExportFound, RoslynCSharpFormatName, this.projectKey);
```

Comment:
The SonarLint provider called two profile exporters, one to fetch the ruleset and the other to fetch the rule configuration file. The new exporter returns one XML file containing all of the configuration data, which this class deserializes and processes i.e. extracts the ruleset and writes it to disc; extracts and writes any additional files; and provisions any NuGet packages.